RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,323 @@
+  shape checking code between CPU and CUDA implementations, as well as
+  with out-of-tree backend implementations
+
+* **Reduce functional/inplace/out boilerplate**: avoid having to write
+  foo, foo_, foo_out variants for every function; make it harder to
+  forget to add foo_out variant when it is appropriate
+
+* **Be able to run forward shape computation without kernel**: so
+  compilers and static runtimes can easily compute sizes of all elements
+  in the framework without having to actually run the computation
+
+* **Provide entry point for static runtime**: provide public API for
+  accessing operators directly, bypassing output allocation, shape
+  checking, device guards.
+
+* **Bypass dispatcher overhead**: in high performance cases, e.g., core
+  framework operators, reduce the number of redispatches to improve
+  performance
+
+* **Unify with version counter bumps and view metadata tracking**: these
+  are currently done in autograd generated code but must be performed
+  unconditionally even when autograd is disabled. This logic to be
+  incorporated with logic here
+
+* **Work with mobile selective build**: mobile selective build is
+  implemented in codegen, and implementation strategy must be compatible
+  with design constraints in mobile
+
+## Non-goals
+
+* **Directly handle TensorIterator**: TensorIterator is a uniquely
+  complicated mechanism, and we think it should be designed for on its
+  own.  However, we think that this is an important case and should
+  be handled sooner rather than later.
+
+* **Be zero runtime cost all the time**: We are willing to give up some
+  runtime efficiency for cleaner code. We are willing to give up
+  efficiency *by default* for out of tree implementers, unless they opt
+  in to higher performance. There is always an escape hatch to be high
+  performance if absolutely necessary
+
+* **No codegen**: As long as it is possible to implement things out of
+  tree (at the cost of human understandable boilerplate), codegen is a
+  reasonable strategy for implementing features we need
+
+* **Major format change to native_functions.yaml**: We could do this. We
+  are choosing not to in order to reduce the degrees of freedom in what
+  changes we make to `native_functions.yaml`.
+
+## Summary
+
+* Core
+
+    * Define a new API for shape checking functions. Shape checking
+      functions are called from generated code that implements
+      functional/inplace/out variants of functions.
+
+    * Define a new API for kernels. Static kernel API does not do shape
+      checking, output is always preallocated at correct size. This API
+      is public and a suitable entrypoint for static runtime. Static
+      kernels are called from generated code like above.
+
+    * Code generate meta functions from shape checking functions to
+      provide public API for running shape computations without any
+      kernel.
+
+    * Generated code is augmented to do version counter bumps and view
```

Comment:
This means we are using this mechanism for all ops, even ops that don't have out or inplace variants, right? Otherwise those ops wouldn't get device guards or version counter bumps.

Reply:
Yeah, this is an orthogonal change that we can do with or without structured kernels.